### DISCUSSION TITLE
Incorrect DB config, or could not connect to DB PDOException: SQLSTATE[HY000] [2002] No such file or directory

### DISCUSSION THREAD

**Discussion by LinusYeung** [YOURLS/YOURLS#3643] [Q&A]
Another issue encountered.
I tried to change the host to 127.0.0.1, it doesn't help.
I have write a simple program to connect the DB by PDO object, it works and it could retrieve data from the DB.
I m not too sure what is missing so that it cost No such file or directory appeared

**Comment by logusivam** [comment]
Since `127.0.0.1` didn't fix it and your own PDO script connects successfully, the problem is not your database server — MySQL is running fine and is reachable. The difference is in **how** your working PDO script connects versus how YOURLS connects. Identifying that difference is what will fix it.

---

## Step 1 — Compare your working PDO script against YOURLS's connection exactly

In your working script, print out every connection parameter you're using:

```php
<?php
$host    = '127.0.0.1';   // or whatever you used
$port    = 3306;           // are you specifying a port explicitly?
$dbname  = 'yourls';
$user    = 'yourls_user';
$pass    = 'yourpassword';
$charset = 'utf8mb4';

$dsn = "mysql:host=$host;port=$port;dbname=$dbname;charset=$charset";

$options = [
    PDO::ATTR_ERRMODE => PDO::ERRMODE_EXCEPTION,
];

$pdo = new PDO( $dsn, $user, $pass, $options );
echo "Connected successfully";
echo "\nHost info: " . $pdo->getAttribute( PDO::ATTR_CONNECTION_STATUS );
```

Run this and note exactly which `$host`, `$port`, and DSN format makes it work. Then check your `user/config.php` and confirm YOURLS is using the exact same values.

---

## Step 2 — Check if a custom port or socket is needed

The `[2002] No such file or directory` error specifically means PHP is trying to use a **Unix socket** instead of a TCP connection — even when the host is set to `127.0.0.1`. This can happen when:

**Your MySQL is running on a non-standard port.** If MySQL is not on port 3306, YOURLS doesn't know this unless you specify the port in `YOURLS_DB_HOST`. Try:

```php
// In user/config.php — append the port directly to the host
define( 'YOURLS_DB_HOST', '127.0.0.1:3366' );  // replace 3366 with your actual port
```

Find your actual MySQL port by running:

```bash
mysqladmin variables | grep port
# or
mysql -u root -p -e "SHOW VARIABLES LIKE 'port';"
```

**Your MySQL is configured to only accept socket connections.** Check your MySQL config:

```bash
cat /etc/mysql/mysql.conf.d/mysqld.cnf | grep -E "bind|socket|skip-networking"
```

If you see `skip-networking` or `bind-address = 127.0.0.1` with a socket-only config, add the socket path explicitly to YOURLS:

```php
define( 'YOURLS_DB_HOST', '127.0.0.1:/var/run/mysqld/mysqld.sock' );
```

Find your actual socket path:

```bash
mysqladmin variables | grep socket
# Common paths:
# /var/run/mysqld/mysqld.sock   (Ubuntu/Debian)
# /tmp/mysql.sock               (macOS, some shared hosts)
# /var/lib/mysql/mysql.sock     (CentOS/RHEL)
```

---

## Step 3 — Check which PHP version and which php.ini YOURLS is using

Your working PDO script and YOURLS may be running under **different PHP versions** with different `php.ini` settings — especially common on shared hosting where the CLI PHP and the web server PHP are separate.

Check which PHP your web server is using:

```php
// Create a temporary file: yourls-root/phpinfo-test.php
<?php phpinfo();
```

Visit `https://yourdomain.com/phpinfo-test.php` in your browser and look for:

- **PHP Version** at the top
- **pdo_mysql.default_socket** under the PDO section
- **mysqli.default_socket** for comparison

Then delete `phpinfo-test.php` immediately after checking.

Compare the `pdo_mysql.default_socket` value in the browser output against the actual socket path from Step 2. If they don't match, that is your problem — YOURLS's PHP is looking for a socket file at the wrong path.

Fix it by adding to the correct `php.ini` (the one shown in phpinfo under **Loaded Configuration File**):

```ini
pdo_mysql.default_socket = /var/run/mysqld/mysqld.sock
```

Then restart your web server:

```bash
sudo systemctl restart apache2   # or nginx / php-fpm depending on your setup
```

---

## Step 4 — Check if YOURLS's DB user has TCP connection permission

MySQL grants are per host. Your DB user may have been granted access only via socket (shown as `localhost` in MySQL grants) but not via TCP (`127.0.0.1`):

```sql
-- Run this in MySQL as root
SELECT user, host FROM mysql.user WHERE user = 'your_yourls_db_user';
```

If you see only `localhost` (socket) and not `127.0.0.1` (TCP), grant TCP access:

```sql
GRANT ALL PRIVILEGES ON yourls_db.* TO 'your_yourls_db_user'@'127.0.0.1' IDENTIFIED BY 'yourpassword';
FLUSH PRIVILEGES;
```

---

## Step 5 — Check if a Docker or container network is involved

If YOURLS is running inside Docker (even the official image), `127.0.0.1` inside the container does **not** point to your MySQL server — it points to the container itself. The correct host in Docker is either:

- The Docker service name if MySQL is also in Docker (e.g. `db` or `mysql` as defined in `docker-compose.yml`)
- The Docker host gateway IP (`172.17.0.1` by default, or use `host.docker.internal` on Docker Desktop)

```php
// If MySQL is a separate Docker service named "db"
define( 'YOURLS_DB_HOST', 'db' );

// If MySQL is on the Docker host machine
define( 'YOURLS_DB_HOST', 'host.docker.internal' );  // Docker Desktop (Mac/Windows)
define( 'YOURLS_DB_HOST', '172.17.0.1' );            // Docker on Linux
```

---

## Diagnostic summary — work through in order

| Step | What to check | What to look for |
|------|--------------|-----------------|
| 1 | Print your working PDO connection params | Exact host, port, DSN format that works |
| 2 | Check MySQL port and socket path | Non-standard port or socket path mismatch |
| 3 | Check `phpinfo()` in browser | `pdo_mysql.default_socket` value vs actual socket path |
| 4 | Check MySQL user grants | `127.0.0.1` grant may be missing, only `localhost` exists |
| 5 | Check if Docker is involved | `127.0.0.1` means the container itself, not the host |

The fact that your own PDO script works is the most useful clue — post the exact DSN and connection options from that working script alongside your YOURLS `YOURLS_DB_HOST` value, and the difference between the two will point directly at the fix.

Hope this helps — if it answers your question, would you mind clicking **"Mark as answer"** so others searching for the same thing can find it quickly?
